app/src/shared/components/UndoContextProvider/index.jsx

```diff
@@ -204,15 +207,28 @@ class UndoContextProvider extends Component<Props, State> {
      * Reset to initialState
      */
 
+    setInitialState = (fn: Function, done: Function) => {
+        const p: Promise<void> = new Promise((resolve) => (
+            this.setState(({ initialState }) => {
+                const nextInitialState = (typeof fn === 'function') ? fn(initialState) : initialState
+                return {
+                    initialState: nextInitialState,
+                }
+            }, resolve)
+        ))
+        p.then(done)
+        return p
```

Comment:
Not trying to be difficult but this still allows `history[0]` and `initialState` to get out of sync. 

e.g. what does `initialState` mean in the context of an undo stack if it's not in the `history` and you can't actually undo to that state?  It seems like a totally separate value that only has meaning *outside* the undo stack. i.e. consider no more `initialState` value, just read and write to history item `0`, or cache the value in the component that needs it?

Reply:
@timoxley Yes, you are right that `state[0]` and `initialState` can get out of sync but that's kind of the point as I see it. Maybe you had a different idea in mind when you started this component but the `initialState` is not used in the editor at the moment so I am reusing it for my purposes here. 

But as it is now, if `state[0]` is the initial state then we don't need `initialState`, right?

The `state[0]` can be replaced so the original value is lost in that case. And if there is no `initialState` what does `reset` do?

We could take out the `initialValue` from here just use the undo stack and then keep the original value in another provider. But this way we don't need to two contexts/providers for it.